packages/@react-aria/textfield/src/useFormattedInput.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {RefObject, useEffect, useRef} from 'react';
+
+function supportsNativeBeforeInputEvent() {
+  return typeof window !== 'undefined' &&
+    window.InputEvent &&
+    // @ts-ignore
+    typeof InputEvent.prototype.getTargetRanges === 'function';
+}
+
+export function useFormattedInput(props, state: {validate: (val: string) => boolean, setInputValue: (val: string) => void}, inputRef: RefObject<HTMLInputElement>) {
```

Comment:
I think we should name this `useFormattedTextField`. You also need types for these. I think `props` should be the same as `useTextField` accepts, and you should move the type for `state` into an interface. 

Another idea I had would be for `useFormattedTextField` to call `useTextField` internally rather than the user needing to take the returned props and pass them to `useTextField` themselves. What do you think?

Reply:
that's a good idea i think, lemme go try that out